### ISSUE TITLE
Choosing installation directory and other installation options

### ISSUE THREAD

**Issue by BramVanroy** [atom/atom#7504]
So far I really like Atom 1.0 - it's got a long way to go, but it will get there. We needed an editor like this.

Anyway: the only thing bothering me was how quickly it installed itself (Windows 8.1, 64 bit). Even though that might sound like a *good*¨thing, it isn't. When I double-click the installer I get no options, no information whatsoever. Only a screen "Atom is installing and will launch when ready." For some people this might be the best approach, but for a whole lot of others it isn't.

Some options would be great. I am thinking: file association, installation directory, packages to include with installation (I for one don't need all packages that are included on default install) and so on. I could understand that file association and package control aren't included (though I'd really like that) but why not the ability to choose your own installation directory?

**Comment by batjko** [comment]
Since this seems to include what you wrote under #7503, can you close that one, please?

In general this request has been made before, though I seem to be unable to find the issue for it. Someone else will know and reference it.

**Comment by BramVanroy** [comment]
Hm, I have no idea how that duplicate happened. I closed it. 

I also searched for this issue before but surprisingly couldn't find it. So I thought I'd create one myself. :-)

**Comment by batjko** [comment]
Yea no problem. There are people looking after issues that will remember any past tickets about this and mention it if applicable.

Here's one for file associations for example:
https://github.com/atom/atom/issues/4893

And here I found a good one on the installation directory question:
https://github.com/atom/atom/issues/5259

Have a look at those.

**Comment by Stian222** [comment]
I downloaded atom for the first time. 
Thing is, i've got a small SSD harddrive and another bigger harddrive that I install most stuff on. So i got really really irritated when the software didn't let me choose install path by default. 

I proceeded to uninstall and reinstall just to check if I'd clicked past a 'choose directory' screen without noticing.

**Comment by Zireael07** [comment]
@Stian222: There is a portable version but I can't get instructions on how to get it installed to a different drive. We seem to be in the same boat :D

**Comment by Carighan** [comment]
I would also like to add a voice to this. I am in a similar situation as @Stian222 , I have a SSD as my C-drive and I would like Atom, large as it is (>300MB after install with just one version kept in the folder) to not be on that drive.

However, there is no option for it.

Likewise, context menu entry and file associations are not an option, and the context menu isn't even removed on uninstall.

**Comment by damieng** [comment]
The belief here is that the installer should be unobtrusive and meets the needs of the majority of people.  

You can download atom-windows.zip from https://github.com/atom/atom/releases/latest and unpack where you want it. With the portable app support it will just work.

We're also considering moving the shell integration options into settings so you could still easily enable/disable that.

**Comment by J0celyn** [comment]
My belief is that an installer should **always** ask me where it should install the program.  
I am the user, I know better than the installer or the Atom programmer team where I want my files to be on my computer.  
I find the current way the installer does it very unpleasant and think there should be a dialog for the user to specify the install directory, as almost all other installers do it on Windows.
And if I decide to install it in "Program Files" and get UAC dialogs when updating, then that's my problem, not yours.

**Comment by Nexxado** [comment]
extremely annoying that there's no option to choose installation directory.

**Comment by damieng** [comment]
If you want to chose a specific folder I'd recommend downloading atom-windows.zip and extracting to the folder of your choice.

If you want it to also create an icon, file association and put it in the path open a command prompt to the folder you extracted and type;

```
atom --squirrel-install
```
which will go ahead and perform those steps for you with the correct location you extracted to.

**Comment by lovebdsobuj** [comment]
Please give us this feature,
> Choosing installation directory and other installation options.

_Automatic install on UserProfile really it's ugly things anyway._

**Comment by lovebdsobuj** [comment]
@damieng, 
I run this command line, but nothing happens `atom --squirrel-install`

**Atom Version:** `1.11.2`
**Platform:** `Windows 10 / 32 bit`

**Comment by damieng** [comment]
@lovebdsobuj Installing into c:\program files means you would need UAC elevation on install and every update.

What error did you get from the `atom --squirrel-install` ?

**Comment by lovebdsobuj** [comment]
Maybe it's work. At first I had installed Atom on my user profile. Somehow the installation dir wiped automatically. Later I download `atom-windows.zip` and extarcted it to other drive and it's works.

**Comment by Ninad998** [comment]
@damieng , but doesn't not asking for location defeat the purpose of it being a "hackable text editor".
How do I get the freedom I want with the application if all the things I can do with it involve things only after the installation?
And I get that the issue with the permissions is a big problem, but doesn't giving the user the choice means its his/her hassle and not the editor's?
I would really like to see a version that has a default install location, but also gives the freedom of choosing location and file association in the installer itself.

But other than the above I think this is a far better and lighter editor than most I have used. Good job @team

**Comment by rm-v** [comment]
Hi, 
atom v.1.12.1
I tried `AtomSetup.exe --machine` - it didn't work
Then I used zip version and used 'atom --squirrel-install' - it did nothing (no error). Also this option isn't listed in 'atom -h'
UAC is not a good argument, since it should be my (user) decision. I use special directory on windows for programs which have some problems with UAC or spaces in path etc. and I think user should be asked about installation path. Otherwise it's  mentality of corporations and arrogant (like in: "we know better") 
No offence intended.
thx

**Comment by damieng** [comment]
@Ninad998 It's hackable in the sense that you can hack on it - customize it etc.  As to it being the user's choice - the reality is people will still file bugs if the auto-updater doesn't work for their known config.

@rm-v atom --squirrel-install is both silent and undocumented in the -h but it should have set the path for you and registered it as a file handler.

It's not a case of we think we know better, there are a number of practical reasons.

The current installer is completely unattended and hands-off and that works for a massive majority of our users.  If we stop to ask for an installation directory that most people won't change that's a lot of inconvenience for many people for the convenience of very few.

Now if that was the only way to install Atom it might still be worth it but Atom unlike many applications is completely happy being just extracted anywhere you like.  Advanced users who like to put apps in specific places (and believe me, I'm one) are more than capable of unzipping a file there.  In fact I'd argue it's faster and easier than going through a setup wizard.

All --squirrel-install actually does is add atom to the path and register it as a file handler. The former is only needed if you want to use it on the command line. The latter can be toggled on as a check-box in Settings > System > Register as a file handler.

**Comment by Ninad998** [comment]
@damieng , I am not trying to say that the way it currently installs is not efficient, I would actually argue that it has one of the best install methods I have seen.

I also agree that the best way of using Atom has been the download zip and extract method.

Though judging from the reading I tried to do for changing the install path, an alternative can also be something along the lines of adding an option in the Settings for changing the location it is saved at, that is, complete the install on the default location and let the user (who wants to change location), find and use the option of changing where it is installed for him/her.

This lets the everyone benefit from the convenience of this awesome installer but also lets people choose the location.

**Comment by Zireael07** [comment]
> Though judging from the reading I tried to do for changing the install path, an alternative can also be something along the lines of adding an option in the Settings for changing the location it is saved at, that is, complete the install on the default location and let the user (who wants to change location), find and use the option of changing where it is installed for him/her.

This would be a nice compromise indeed for those not wanting to install to default location.

**Comment by damieng** [comment]
I'm not sure how easy it is to tell the Windows installation registry about a change of location - if it's at all possible. I'll do some digging.

One thing I have considered is putting a checkbox in System preferences that add/removes that version of Atom to the PATH which would give a visible easy-way to get the same result as the installer.

**Comment by AluminiumTech** [comment]
Have the devs dismissed this or will they allow people to choose the installation path?

**Comment by damieng** [comment]
@AluminiumTech There is no plan to allow users to configure an installation path. Right now the installer is silent and one-click and doing so would break that experience.

You can always unpack the portable ZIP version wherever you like.

**Comment by KayJay89** [comment]
Am I correct to assume that even with the latest version, Atom still leaves around trash in the registry and the .atom folder with preferences when trying to uninstall it completely? 

I have been avoiding using Atom because of this particularity given in case something goes wrong, a simple reinstall does not tend to resolve the problem as result of these things being left upon uninstalling Atom.

**Comment by damieng** [comment]
@KayJay89 Atom by default only adds a single entry to the registry to mark itself as being capable of opening files. You can turn this off in Settings > System.  There are some optional entries for enabling it in the Explorer right-menu button. This is the only registry use Atom does and all are removed during uninstall.

Atom never needs to be reinstalled to address a problem - if you wish to reset Atom the approach is simply to delete the %userprofile%\.atom folder which will be recreated with all fresh preferences and state.

If you have any further concerns please use our https://discuss.atom.io or raise a new issue rather than try to repurpose an unrelated issue - this avoids inconveniencing users with unrelated notifications.

**Comment by RedFantom** [comment]
For anyone who is still looking for a custom installation directory option, I have repacked the Atom portable ZIP file into an installer using Inno Setup. You can find the script [here](https://www.github.com/RedFantom/atom-windows-installer) and the installer [in the releases section](https://www.github.com/RedFantom/atom-windows-installer/releases). I created this mainly for myself, as I was annoyed, but I thought it'd be nice to share it here as this is the landing page when googling changing the installation directory :) . Updates probably don't work, but I haven't tested them yet.

**Comment by RedFantom** [comment]
@labibzeidan Is this with my installer for Atom Windows? Because then you could open an issue in my repository. Is it in any other way related to this issue, then you're fine, but going on the little information you give, this sounds like a completely separate program, and then you should open a new issue.

**Comment by naughtuple** [comment]
@RedFantom THANK YOU! Got the context-menu option and everything working, even though I installed it to my portable drive :) Any idea how to get the apm commands to work?

**Comment by RedFantom** [comment]
I'm very glad I could help @naughtuple :) . However, since this conversation will probably disturb the Atom developers, I suggest that we look into the apm commands in an issue in my repository ;) .

**Comment by nguyenlekhtn** [comment]
After one year, this issue still wasn't solved. That's weird
All my program that I have installed has options to choose installation path. I don't think it's difficult

**Comment by bensku** [comment]
Atom is good editor, but this is really annoying. Installation without user intervention beyond opening the installer has a few issues:
* Adminstator/root access to computer is not guaranteed nor does the download page point to Github releases page
* In some environments, default installation location might not have much space - and if OS drive gets out of space, weird things may begin to happen
* Installer does not ask user if it is ok to use X amount of space that Atom needs to be installed, so even if there is enough space it doesn't exactly help me trust app vendor - trying to hide application size?

When either of these happens, your silent and intuitive installer is suddenly an useless (and potentially harmful) one.

**Comment by pha-Z** [comment]
Since c drive mostly is (and most certainly will be in future too) SSD while d drive being HDD, the demand for a custom installation directory feature will certainly rise :/

**Comment by Rinoka** [comment]
@nguyenlekhtn If you read, you would see that the reason is not difficulty - it is an issue of their installer being for a basic user, while their Advanced option is using the portable functionality to unzip.

@bensku: Agreed; @damieng, if portable extraction is the Advanced option, please put a download link for the latest stable ZIP onto atom.io. Installers serve the purpose of informing a user if they can fit the data the program wants to store; your Installer doesn't do that, and while it doesn't use a lot of memory, it takes up space in AppData, a storage that is fundamentally difficult to move and inhibits users with a OS SSD and a storage side drive, especially because Atom stores multiple versions of itself.

However, you could also just make an advanced installer offered on the Atom.io website, advertise it as advanced, and only hinder users that choose to use it... after all, most users probably would like the one-click experience more, but it's kind of shit for people who NEED to pay attention to where things go.

**Comment by damieng** [comment]
@Rinoka unfortunately the amount of work to do this is non-trivial. A second installer would need to:

1. Be able to be built on our CI systems in an automated manner
2. Be able to perform the update duties using existing Squirrel infrastructure
3. Allow binaries to be signed both before and after the installer creation

We are a small team and have to choose where to spend our resources.  Between the basic installer suiting the majority of our users needs and the ZIP file being the flexible option for advanced users you're left with a small number of users requiring a large amount of effort.

The reality is unless the Squirrel installer system supports this then this we're just never going to be able to make the justification for this given all the other features and fixes that affect a greater number of users.

**Comment by wuya666wuya** [comment]
I'd really love a command line option or environment variable to set the program installation location for the Atom installer. I don't need an interactive UI to select the destination folder, but I think it's really an annoying issue to not able to choose the installation folder for the Atom installer. Unzipping and running --squirrel-install is just not elegant enough for Windows as it will not take care of the system registries nor menu shortcuts nor uninstallation handlers. Also for someone used to utilize the Cortana to launch applications quickly, it will not index the Atom editor if it's not unpacked into some common library folders. 

I think a simple command line switch like AtomSetup-x64.exe /dest "D:\Utilities\Editors\" would be fantastic, and it should be very easy to implement.

**Comment by damieng** [comment]
@wuya666wuya We use the Squirrel installer system.  Squirrel would need to support that, not Atom.

**Comment by Ninad998** [comment]
Well, clearly Squirrel could care less:
https://github.com/Squirrel/Squirrel.Windows/issues/1002
I think this suggests a flaw in the choice of installer

**Comment by dizzy-egg** [comment]
How incredibly rich that Atom, which advertises itself in the very strapline of its website as being "a tool you can customize to do anything" cannot even be installed to a location the user wants.

The "world's most customizable text editor", and _literally_ the first thing it does is remove a choice from the user. Let the irony of that really sink in for a moment.

Seriously. This is the worst possible installation choice that could have been made. As usual, it's a case of _developers knows best™_.

**Comment by lee-dohm** [comment]
@dizzy-egg your comment was deleted as a violation of the [Atom Code of Conduct](https://github.com/atom/atom/blob/master/CODE_OF_CONDUCT.md) as it was insulting or derogatory. You may consider this an official warning.
